File: couchbase-entity/src/main/java/com/kaufland/generation/CblDefaultGeneration.java
package com.kaufland.generation;

import com.kaufland.model.entity.BaseEntityHolder;
import com.kaufland.model.field.CblFieldHolder;
import com.kaufland.util.TypeUtil;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.FieldSpec;
import com.thoughtworks.qdox.model.JavaField;

import javax.lang.model.element.Modifier;
import javax.lang.model.type.TypeMirror;

public class CblDefaultGeneration {

    public static FieldSpec field() {
        return FieldSpec.builder(TypeUtil.createMapStringObject(), "mDocDefaults", Modifier.PRIVATE, Modifier.STATIC).
                initializer(" new $T()", TypeUtil.createHashMapStringObject()).
                build();
    }

    public static CodeBlock staticInitialiser(BaseEntityHolder holder) {
        CodeBlock.Builder builder = CodeBlock.builder();
        for (CblFieldHolder fieldHolder : holder.getFields()) {

            if (fieldHolder.isDefault()) {
                builder.addStatement("mDocDefaults.put($N, " + getConvertedValue(fieldHolder.getTypeMirror(), fieldHolder.getDefaultValue())+")", fieldHolder.getConstantName());
            }
        }
        return builder.build();
    }

    private static String getConvertedValue(TypeMirror clazz, String value) {

        if (clazz.toString().equals(String.class.getCanonicalName())) {
            return "\"" + value + "\"";
        }
        return value;
    }

}

File: couchbase-entity/src/main/java/com/kaufland/generation/CblDefaultGeneration.kt
package com.kaufland.generation

import com.kaufland.javaToKotlinType
import com.kaufland.model.entity.BaseEntityHolder
import com.kaufland.util.ConversionUtil
import com.kaufland.util.TypeUtil
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.KModifier
import com.squareup.kotlinpoet.asTypeName
import javax.lang.model.type.TypeMirror

object CblDefaultGeneration {

    fun addDefaults(holder: BaseEntityHolder, useNullableMap : Boolean): FunSpec {

        val type = if (useNullableMap) TypeUtil.mutableMapStringAnyNullable() else TypeUtil.mutableMapStringAny()
        val typeConversionReturnType = if (useNullableMap) TypeUtil.anyNullable() else TypeUtil.any()

        val builder = FunSpec.builder("addDefaults").addModifiers(KModifier.PRIVATE).addParameter( "map", type)

        for (fieldHolder in holder.fields.values) {

            if (fieldHolder.isDefault) {
                builder.addStatement("map.put(%N, " +  fieldHolder.ensureType(typeConversionReturnType, ConversionUtil.convertStringToDesiredFormat(fieldHolder.typeMirror, fieldHolder.defaultValue)) + "!!)", fieldHolder.constantName)
            }
        }
        return builder.build()
    }

    fun addAddCall(nameOfMap: String): CodeBlock {
        return CodeBlock.builder().addStatement("addDefaults(%N)", nameOfMap).build()
    }

}
